### ISSUE TITLE
Apply type constraints on generic types for objects with generic types even when the generic type is unknown.

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#2460]
**Is your feature request related to a problem? Please describe.**
Suppose you have a function like so:
```py
from typing import Iterable, TypeVar

ItemT = TypeVar("ItemT", bound=Any)

def some_func(some_iter: Iterable[ItemT]) -> Any:
    def exec(item: ItemT) -> Any:
        ...

    for item in some_iter:
        return exec(item)  # Only `item` is a valid argument here.
```
here the `exec` function will only ever accept a variable of type `ItemT`. However if we setup the same function into a class that mimics `some_func`s purpose then we can run into situations where this relationship between the `Iterable` and `exec` is not preserved:

```py
from typing import Iterator, TypeVar

ItemT = TypeVar("ItemT", bound=Any)

class SomeClass(Generic[ItemT]):
    def make_iter(self) -> Iterator[ItemT]:
        ...

    def exec(self, item: ItemT) -> Any:
        ...
```

Here we have a class `SomeClass` that is generic over the same type `ItemT`. However if we were to use this class like so:

```py
some_obj = SomeClass()

for item in some_obj.make_iter():
    some_obj.exec("hi")  # No error because `item` is `Unknown`.
    some_obj.exec(item)

```
We see that we didn't get any error on `some_obj.exec("hi")`, despite this violating the relationship between `ItemT`, `exec` and `make_iter` inside `SomeClass`

Because we didn't define what `SomeClass` is generic over, `ItemT` gets type `Unknown`, and to my understanding pyright will not apply any type checking with `Unknown` types, hence we no longer get any type checking on `some_obj.exec`, despite the class being defined as having this strict relationship. 

Now in this case the obvious solution is just to add type arguments when creating `SomeClass`. However, when I define the type relationships in a class I would expect those to always be obeyed from a typing perspective, regardless of whether or not generic types are supplied when the object is created.

**Describe the solution you'd like**
Maybe in these cases the type of `item` should not be `Unknown` but something like `ItemT@SomeClass`, preserving the relationship that `ItemT` has inside `SomeClass`.

This would potentially also improve the situation where a method takes a generic type as argument, but the object has unknown generic types, resulting in no type checking on the argument. Better explained with code sample:

```py
from typing import TypeVar

T = TypeVar("T", bound=int) # Note `T` is bound to `int`

class OtherClass(Generic[T]):
    def other_func(self, some_var: T) -> int:
        ...

OtherClass().other_func("hello") # No error because `T` is `Unknown`
```

Despite `T` being bound to `int`, I'm able to pass a string into `other_func` because no type checking is applied. 

Without knowing pyright internals, my intuition tells me that both cases would be fixed if pyright would default back to the `TypeVar` of the generic type when no type arguments are supplied (this might have side effects I may not be aware of). Apologies if this is a separate issue, but they seem linked to me, let me know if I should create a separate issue for this case.

**Additional context**
mypy does not handle these cases either, and also just defaults to their equivalent in `Any`.

**Comment by erictraut** [comment]
In your first example, you've bound a TypeVar to `Any`. You shouldn't do this because it's an overly-complex equivalent of `Any`. If you intend `Any`, just use `Any`. Otherwise use a regular TypeVar without a bound type (which effectively binds it to `object`).

> Because we didn't define what SomeClass is generic over, ItemT gets type Unknown, and to my understanding pyright will not apply any type checking with Unknown types...

You're correct in that assumption. If your constructor doesn't provide sufficient information to specialize the class, you need to specialize the class explicitly at construction time, like this:

```python
some_obj = SomeClass[int]()
```

> Maybe in these cases the type of item should not be Unknown but something like ItemT@SomeClass, preserving the relationship that ItemT has inside SomeClass.

The type of `ItemT` needs to be established at the time the object is constructed. I think it's better to explicitly generate an `Unknown` in this case so the error is flagged in strict mode. This allows you to detect (and correct) the error by explicitly specializing the class at construction time.

If you enable strict mode, pyright will generate errors when you attempt to use a value with an Unknown or partially-Unknown type.
